### PR TITLE
[DEV 3964] Upgrade Pandas to 2.2.2

### PR DESCRIPTION
**Pull request by nonibansal**

### CODE REVIEW

File: fennel/client_tests/branches/test_clone.py [hunk 1]
```diff
@@ -171,7 +171,13 @@ def test_clone_errors(client):
     # can not clone to an existing branch
     with pytest.raises(Exception) as e:
         client.clone_branch("main", from_branch="test-branch")
-    assert str(e.value) == "Branch `main` already exists"
+    if client.is_integration_client():
+        assert (
+            str(e.value)
+            == "Server returned: 500, can not clone from branch `random`: does not exist"
+        )
```
Comment (aditya-nambiar) [inline fennel/client_tests/branches/test_clone.py:178]: This error does not look right to me, random branch is not present on line 173 ?

File: fennel/datasets/datasets.py [hunk 2]
```diff
@@ -1203,7 +1205,7 @@ def lookup(
                 res[col] = res[col].apply(
                     lambda x: parse_json(type_annotation, x)
                 )
-            return res.replace({np.nan: None}), found
+            return res, found
```
Comment (aditya-nambiar) [inline fennel/datasets/datasets.py:1208]: very cool. So now the backend itself returns pd.NA ?

File: fennel/testing/test_utils.py [hunk 3]
```diff
@@ -167,6 +257,42 @@ def proto_to_dtype(proto_dtype) -> str:
...
+                f"Failed to cast column `{f.name}` of type `{proto_to_dtype(f.dtype)}` to arrow dtype. Error: {e}"
+            )
+    return df
+
+
+def cast_df_to_pandas_dtype(
+    df: pd.DataFrame, fields: List[Field]
+) -> pd.DataFrame:
+    """
+    This helper function casts Pandas DatFrame columns to arrow dtype using list of Field proto.
```
Comment (aditya-nambiar) [inline fennel/testing/test_utils.py:282]: Both function have the same description. This one is incorrect i believe ?

File: fennel/client/client.py [hunk 4]
```diff
@@ -813,6 +844,21 @@ def _post(
         check_response(response)
         return response
 
+    @staticmethod
+    def _parse_dataframe(
```
Comment (aditya-nambiar) [inline fennel/client/client.py:848]: Can you add a description to this function.
Also where is it supposed to be used ?

File: fennel/internal_lib/schema/schema.py [hunk 5]
```diff
@@ -224,6 +226,45 @@ def get_python_type_from_pd(type):
     return type
 
 
+def convert_dtype_to_arrow_type(dtype: schema_proto.DataType) -> pa.DataType:
```
Comment (aditya-nambiar) [inline fennel/internal_lib/schema/schema.py:229]: are there tests for all the conversion functions including this?

File: pyproject.toml [hunk 6]
```diff
@@ -7,8 +7,8 @@ packages = [{ include = "fennel" }]
 readme = "README.md"
 
 [tool.poetry.dependencies]
-python = "^3.8"
-pandas = "^1.0.0"
+python = "^3.9"
```
Comment (aditya-nambiar) [inline pyproject.toml:10]: Bump up the version of the client

File: fennel/client_tests/test_dataset.py [hunk 7]
```diff
@@ -321,7 +321,7 @@ def test_log_to_dataset(self, client):
                 response = client.log("fennel_webhook", "UserInfoDataset", df)
             assert (
                 str(e.value)
-                == """Schema validation failed during data insertion to `UserInfoDataset`: Failed to cast data logged to column `age` of type `optional(int)`: Unable to parse string "32yrs" at position 0"""
+                == """Schema validation failed during data insertion to `UserInfoDataset`: Failed to cast data logged to column `age` of type `optional(int)`: Expected bytes, got a 'int' object"""
```
[the remark this answers is not in the document]
Comment (aditya-nambiar) [reply fennel/client_tests/test_dataset.py:489]: I dont get the error message, you expect an int right ? but incorrectly received a string ?

File: fennel/testing/test_utils.py
Comment (aditya-nambiar) [inline fennel/testing/test_utils.py:1]: are there tests for these utils ?

File: pyproject.toml (hunk 6 above)
Comment (mohitreddy1996) [inline pyproject.toml:10]: should we remove the CI which runs for 3.8 as well? Given we have declared incompatibility for it?

File: fennel/client_tests/test_dataset.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nonibansal) [reply fennel/client_tests/test_dataset.py:489]: Yup we expected an int but got a string, this is some weird error that pandas throw. Any idea on how we can change this?

File: fennel/datasets/datasets.py (hunk 2 above)
Comment (nonibansal) [reply fennel/datasets/datasets.py:1208]: Yup, or we patch that afterwards.

File: pyproject.toml (hunk 6 above)
Comment (aditya-nambiar) [reply pyproject.toml:10]: I have already removed it

File: fennel/client_tests/branches/test_clone.py (hunk 1 above)
Comment (nonibansal) [reply fennel/client_tests/branches/test_clone.py]: changed this.

File: fennel/testing/test_utils.py (hunk 3 above)
Comment (nonibansal) [reply fennel/testing/test_utils.py]: yup changed the description

File: fennel/client/client.py (hunk 4 above)
Comment (nonibansal) [reply fennel/client/client.py:848]: added

File: fennel/internal_lib/schema/schema.py (hunk 5 above)
Comment (nonibansal) [reply fennel/internal_lib/schema/schema.py:229]: added tests

File: fennel/testing/test_utils.py
Comment (nonibansal) [reply fennel/testing/test_utils.py:1]: added

File: pyproject.toml (hunk 6 above)
Comment (nonibansal) [reply pyproject.toml:10]: bumped

**Review by aditya-nambiar — APPROVED** [review]

**Pull request merged**